File: neetcode150/10_backtracking/0079_word_search.cpp
/*
 * =====================================================================
 * LeetCode 79  : Word Search                                         (Medium)
 * https://leetcode.com/problems/word-search/
 * Category   : Backtracking on a grid
 * ---------------------------------------------------------------------
 * PROBLEM
 *   Does `word` exist as a path of ADJACENT cells (up/down/left/right)
 *   in an m×n board of letters? A cell may be used at most once.
 * ---------------------------------------------------------------------
 * INTUITION
 *   Try every cell as the start; DFS match character by character,
 *   marking visited cells; un-mark when the branch fails (backtrack).
 * ---------------------------------------------------------------------
 * APPROACH 1 · DFS with in-place marking (best)
 *   dfs(i, j, k): board[i][j] must equal word[k]; mark with a sentinel
 *   ('#' or '0'), recurse 4 neighbours for k+1, restore the letter.
 * APPROACH 2 · visited[][] bitmap
 *   Same search but keep a separate bool grid (clearer, 1 extra array).
 * ---------------------------------------------------------------------
 * DEEP DIVE · Sentinel marking — the 1-byte state machine
 *   Instead of a visited[] array we overwrite the cell with a value
 *   that can never match a letter ('\0' or '#'), then restore it on
 *   unwind. This makes "used" implicit in the board itself — no
 *   allocation, perfect cache locality, and the restoration is forced
 *   (any buggy path that forgets to restore corrupts the board, which
 *   makes errors loud). Worst case with all cells matching the first
 *   letter: O(m·n·3^k) (3 neighbours after the first step).
 * ---------------------------------------------------------------------
 * TIME COMPLEXITY : O(m·n·3^L) worst, L = word length.
 * MEMORY COMPLEXITY: O(L) recursion depth.
 * =====================================================================
 */
#include <iostream>
#include <vector>
#include <string>
#include <chrono>
#include <sys/resource.h>

class Solution_1 {
    std::vector<std::vector<char>>* b_;
    std::string* w_;
    int m_, n_;
    bool dfs(int i, int j, int k) {
        if (k == static_cast<int>(w_->size())) return true;
        if (i < 0 || j < 0 || i >= m_ || j >= n_) return false;
        if ((*b_)[i][j] != (*w_)[k]) return false;
        char save = (*b_)[i][j];
        (*b_)[i][j] = '\0';                        // mark visited in-place
        bool ok = dfs(i + 1, j, k + 1) || dfs(i - 1, j, k + 1) ||
                  dfs(i, j + 1, k + 1) || dfs(i, j - 1, k + 1);
        (*b_)[i][j] = save;                        // un-mark (backtrack)
        return ok;
    }
public:
    bool exist(std::vector<std::vector<char>>& board, std::string word) {
        b_ = &board;
        w_ = &word;
        m_ = static_cast<int>(board.size());
        n_ = static_cast<int>(board[0].size());
        for (int i = 0; i < m_; ++i)
            for (int j = 0; j < n_; ++j)
                if (dfs(i, j, 0)) return true;
        return false;
    }
};

class Solution_2 {
    std::vector<std::vector<char>> b_;
    std::string w_;
    int m_, n_;
    bool dfs(int i, int j, int k, std::vector<std::vector<bool>>& vis) {
        if (k == static_cast<int>(w_.size())) return true;
        if (i < 0 || j < 0 || i >= m_ || j >= n_ || vis[i][j]) return false;
        if (b_[i][j] != w_[k]) return false;
        vis[i][j] = true;
        bool ok = dfs(i + 1, j, k + 1, vis) || dfs(i - 1, j, k + 1, vis) ||
                  dfs(i, j + 1, k + 1, vis) || dfs(i, j - 1, k + 1, vis);
        vis[i][j] = false;
        return ok;
    }
public:
    bool exist(std::vector<std::vector<char>>& board, std::string word) {
        b_ = board;
        w_ = word;
        m_ = static_cast<int>(board.size());
        n_ = static_cast<int>(board[0].size());
        std::vector<std::vector<bool>> vis(m_, std::vector<bool>(n_, false));
        for (int i = 0; i < m_; ++i)
            for (int j = 0; j < n_; ++j)
                if (dfs(i, j, 0, vis)) return true;
        return false;
    }
};

static long long nowUs() {
    return std::chrono::duration_cast<std::chrono::microseconds>(
        std::chrono::steady_clock::now().time_since_epoch()).count();
}
static long memKb() { struct rusage r{}; getrusage(RUSAGE_SELF, &r); return r.ru_maxrss; }

int main() {
    std::vector<std::vector<char>> board = {
        {'A', 'B', 'C', 'E'}, {'S', 'F', 'C', 'S'}, {'A', 'D', 'E', 'E'}};
    struct Case { std::string w; bool want; };
    const Case cases[] = {{"ABCCED", true}, {"SEE", true}, {"ABCB", false}};
    const long mem0 = memKb();
    auto t0 = nowUs();
    bool allOk = true;
    for (const auto& c : cases) {
        auto b1 = board, b2 = board;
        bool r1 = Solution_1().exist(b1, c.w);
        bool r2 = Solution_2().exist(b2, c.w);
        bool ok = r1 == c.want && r2 == c.want;
        allOk = allOk && ok;
        std::cout << "\"" << c.w << "\" → A1=" << (r1 ? "T" : "F")
                  << " A2=" << (r2 ? "T" : "F") << " (want " << (c.want ? "T" : "F")
                  << ") " << (ok ? "PASS" : "FAIL") << "\n";
    }
    auto t1 = nowUs();
    std::cout << (allOk ? "PASS : all cases" : "FAIL") << " | time: " << (t1 - t0)
              << " us | mem: " << (memKb() - mem0) << " KB\n";
    return 0;
}

File: neetcode150/10_backtracking/0079_word_search.py
"""
=====================================================================
LeetCode 79 : Word Search                                         (Medium)
https://leetcode.com/problems/word-search/
Category   : Backtracking on a grid
---------------------------------------------------------------------
PROBLEM
    Does `word` appear as a path of adjacent cells (up/down/left/
    right), each cell used at most once, in an m×n board?
---------------------------------------------------------------------
INTUITION
    Try every cell as the start; DFS-match the word character by
    character, marking cells visited, unmarking on failure.
---------------------------------------------------------------------
APPROACH 1 · In-place sentinel marking (best)
    Overwrite the matched cell with a sentinel that can never equal a
    letter, restore it after the 4-neighbour recursion.
APPROACH 2 · visited[][] grid
    Same search with an explicit bool matrix.
---------------------------------------------------------------------
DEEP DIVE · The sentinel state machine
    '\0' can never match an input letter, so "used" is encoded in the
    board itself — no allocation, no restore on the success path
    (early return unwinds without restoring, which is safe because we
    already found the answer). Worst case (all cells match the first
    char): O(m·n·3^L).
---------------------------------------------------------------------
TIME COMPLEXITY : O(m·n·3^L) worst.
MEMORY COMPLEXITY: O(L) recursion depth.
=====================================================================
"""


class Solution_1:
    def exist(self, board, word):
        m, n = len(board), len(board[0])

        def dfs(i, j, k):
            if k == len(word):
                return True
            if not (0 <= i < m and 0 <= j < n) or board[i][j] != word[k]:
                return False
            save = board[i][j]
            board[i][j] = "\0"
            found = (dfs(i + 1, j, k + 1) or dfs(i - 1, j, k + 1) or
                     dfs(i, j + 1, k + 1) or dfs(i, j - 1, k + 1))
            board[i][j] = save
            return found

        for i in range(m):
            for j in range(n):
                if dfs(i, j, 0):
                    return True
        return False


class Solution_2:
    def exist(self, board, word):
        m, n = len(board), len(board[0])
        vis = [[False] * n for _ in range(m)]

        def dfs(i, j, k):
            if k == len(word):
                return True
            if not (0 <= i < m and 0 <= j < n) or vis[i][j] or board[i][j] != word[k]:
                return False
            vis[i][j] = True
            found = (dfs(i + 1, j, k + 1) or dfs(i - 1, j, k + 1) or
                     dfs(i, j + 1, k + 1) or dfs(i, j - 1, k + 1))
            vis[i][j] = False
            return found

        for i in range(m):
            for j in range(n):
                if dfs(i, j, 0):
                    return True
        return False


if __name__ == "__main__":
    import time, tracemalloc
    board = [["A", "B", "C", "E"], ["S", "F", "C", "S"], ["A", "D", "E", "E"]]
    cases = [("ABCCED", True), ("SEE", True), ("ABCB", False)]
    tracemalloc.start()
    t0 = time.perf_counter()
    ok = True
    for w, want in cases:
        r1, r2 = Solution_1().exist([row[:] for row in board], w), Solution_2().exist([row[:] for row in board], w)
        ok &= r1 == r2 == want
        print(f'"{w}" -> A1={r1} A2={r2} (want {want}) {"PASS" if r1 == want and r2 == want else "FAIL"}')
    dt = time.perf_counter() - t0
    mem = tracemalloc.get_traced_memory()[1] / 1024
    tracemalloc.stop()
    print(f"{'PASS : all cases' if ok else 'FAIL'} | time: {dt*1e6:.0f} us | mem: {mem:.0f} KB")
